content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+For instance assertions on the duration of specific requests of a web application cannot be run in a meaningful way on a developer laptop which differs from the actual production hardware.
+When run in a virtualized or containerized CI environment, such tests are prone to failures due to concurrent load of other applications and jobs.
+
+This post introduces https://github.com/gunnarmorling/jfrunit[JfrUnit], which offers a fresh angle to this topic by supporting assertions not on metrics like latency/throughput themselves, but on _indirect metrics_ which may impact those.
+Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation, database I/O, or number of executed SQL statements, for a given workload.
+Starting off from a defined base line, future failures of such assertions are indicators for potential performance regressions in an application, as a code change may have introduced higher GC pressure,
+the retrieval of unneccessary data from the database, or SQL problems commonly induced by ORM tools, like N+1 SELECT statements.
+
+<!--more-->
+
+JfrUnit provides means of identifying and analyzing such anomalies in a reliable, environment independent way in standard JUnit tests,
+before they manifest as actual performance regressions in production.
+Test results are independent from wall clock time and thus provide actionable information also when not testing with production-like hardware and data volumes.
+
+This post is the first one in a series of two:
+
+* Part one provides an overview over continuous performance unit testing with JfrUnit,
+based on the examples of asserting expected memory allocation and database I/O
+* The second part will discuss the assertion of the SQL statements executed by an application,
+based on the instrumentation of JDBC APIs using https://wiki.openjdk.java.net/display/jmc/The+JMC+Agent[JMC Agent]
+
+== Getting Started With JfrUnit
+
+JfrUnit is an extension for JUnit 5, which integrates Flight Recorder into unit tests;
+it makes it straight forward to initiate a JFR recording for a given set of event types,
+execute some test routine, and then assert the JFR events which should have been produced.
+
+Here is a basic example of a JfrUnit test:
+
+[source,java]
+----
+@JfrEventTest // <1>
+public class JfrUnitTest {
+
+  public JfrEvents jfrEvents = new JfrEvents();
+
+  @Test
+  @EnableEvent("jdk.GarbageCollection") // <2>
+  @EnableEvent("jdk.ThreadSleep")
+  public void shouldHaveGcAndSleepEvents() throws Exception {
+    System.gc();
+    Thread.sleep(1000);
+
+    jfrEvents.awaitEvents(); // <3>
+
+    ExpectedEvent = event("jdk.GarbageCollection"); // <4>
+    assertThat(jfrEvents).contains(event); 
+    
+    ExpectedEvent = event("jdk.GarbageCollection") // <4>
+        .with("cause", "System.gc()"));
+    assertThat(jfrEvents).contains(event); 
```

Comment:
pretty cool. So the `with("cause"` created a restriction the matcher ? I mean, I take it that this wouldn't fail if there was a naturally occuring GC event accidentally being recorded as well?

Reply:
Yes, exactly. `with()` sets up the given event attribute to be matched against. If there'd be another GC with another cause, this wouldn't cause this assertion to fail. We might think about adding `containsNot()` though, this could be useful actually. Captured in https://github.com/gunnarmorling/jfrunit/issues/8.